### DISCUSSION TITLE
Execute function of API Extension in a flow

### DISCUSSION THREAD

**Discussion by rokdd** [directus/directus#17140] [Archive]
I am very happy that the flow is developing further and further... I want to execute a function from my API extension out of the flow and the easiest seems for me to use a script block and call per http my own endpoint. Perfect would be when I could access the functions directly with nodejs in the code block but I assume the functions are not available there?

Thanks for your help and support!

**Comment by paescuj** [comment]
Is it a function that somehow depends on the context given when executed as an API extension?
In that case I think calling it via an HTTP request indeed makes the most sense.

If it's more like a shared function and you just want to prevent duplication, you might want to store it in a separate package which you can then
- install and expose to flows via [`FLOWS_EXEC_ALLOWED_MODULES` ](https://docs.directus.io/self-hosted/config-options.html#security)
- depend on and use in your API extension

Possibly, it could also be an applicable use case for a [custom operation](https://docs.directus.io/extensions/operations.html).

**Reply by rokdd** [reply]
Well the problem is for the first solution to HTTP Request to keep the user logged in and use his permissions.

I tried to to put my extensions in [FLOWS_EXEC_ALLOWED_MODULES](https://docs.directus.io/self-hosted/config-options.html#security) but the script block is not working :-( and I guess because it is running isolated in a sandbox it might be also difficult to reuse permissions.

So the only clean thing would be to generate another extension for a custom operation which uses the function of the API extension.

How I can find out how it works to "depend on and use in your API extension" .. how I do that? I am installing the plugin when creating the docker image by copying the files into extension folder, but no glue how to depend them?

**Reply by paescuj** [reply]
> Well the problem is for the first solution to HTTP Request to keep the user logged in and use his permissions.

I'm afraid I don't quite understand. Why / Where does that problem occur?

> I tried to to put my extensions in [FLOWS_EXEC_ALLOWED_MODULES](https://docs.directus.io/self-hosted/config-options.html#security) but the script block is not working :-( and I guess because it is running isolated in a sandbox it might be also difficult to reuse permissions.
>
> So the only clean thing would be to generate another extension for a custom operation which uses the function of the API extension.
>
> How I can find out how it works to "depend on and use in your API extension" .. how I do that? I am installing the plugin when creating the docker image by copying the files into extension folder, but no glue how to depend them?

By that I meant that you can create a separate [package](https://nodejs.org/api/packages.html) where you put shared functions in, and then install (`npm install <my-shared-function-package>`) and depend (`import { mySharedFunction } from '<my-shared-function-package>';` on this package inside your extension.
You could then add this package to your Directus installation, add it to `FLOWS_EXEC_ALLOWED_MODULES` and use it in your flows.

Quick example how your Dockerfile could look like:
```Dockerfile
FROM directus/directus:9.22.4

COPY my-shared-function-package/ .
RUN npm install ./my-shared-function-package
```

Please note that these are just a few general ideas / suggestions... Without knowing the background, it's difficult to provide more specific help.

**Reply by rokdd** [reply]
> > Well the problem is for the first solution to HTTP Request to keep the user logged in and use his permissions.
> 
> I'm afraid I don't quite understand. Why / Where does that problem occur?

I meant the operation "Request Webhook / URL". So after I call my API with the URL I am not able to detect the user anymore in my code because it is like anonymous.

> > I tried to to put my extensions in [FLOWS_EXEC_ALLOWED_MODULES](https://docs.directus.io/self-hosted/config-options.html#security) but the script block is not working :-( and I guess because it is running isolated in a sandbox it might be also difficult to reuse permissions.
> > So the only clean thing would be to generate another extension for a custom operation which uses the function of the API extension.
> > How I can find out how it works to "depend on and use in your API extension" .. how I do that? I am installing the plugin when creating the docker image by copying the files into extension folder, but no glue how to depend them?
> 
> By that I meant that you can create a separate [package](https://nodejs.org/api/packages.html) where you put shared functions in, and then install (`npm install <my-shared-function-package>`) and depend (`import { mySharedFunction } from '<my-shared-function-package>';` on this package inside your extension. You could then add this package to your Directus installation, add it to `FLOWS_EXEC_ALLOWED_MODULES` and use it in your flows.
> 
> Quick example how your Dockerfile could look like:
> 
> ```dockerfile
> FROM directus/directus:9.22.4
> 
> COPY my-shared-function-package/ .
> RUN npm install ./my-shared-function-package
> ```
> 
> Please note that these are just a few general ideas / suggestions... Without knowing the background, it's difficult to provide more specific help.

Ok I got it.. so you meant by package a npm package. I will report later with my success ;-) Thank you so far.

**Reply by paescuj** [reply]
> I meant the operation "Request Webhook / URL". So after I call my API with the URL I am not able to detect the user anymore in my code because it is like anonymous.

Ah, I see.

> Ok I got it.. so you meant by package a npm package. I will report later with my success ;-) Thank you so far.

Great! Much success! 😃
